### PR TITLE
`RSYNC` extension

### PR DESCRIPTION
**Pull request by reidpr**

### CODE REVIEW

**Review by lucaudill — APPROVED** [review]
Sounds like a good plan, I'd say go ahead with it

File: doc/ch-image.rst [hunk 1]
```diff
@@ -1000,11 +1005,660 @@ Features we do not plan to support
...
+Copying files is often simple but has numerous difficult corner cases, e.g.
+when dealing with symbolic or hard links. The standard instruction
+:code:`COPY` deals with many of these corner cases differently from other UNIX
+utilities, lacks complete documentation, and behaves inconsistently between
+different Dockerfile interpreters (e.g., Docker’s legacy builder vs.
+BuildKit), as detailed above. On the other hand, :code:`rsync(1)` is an
+extremely capable, widely used file copy tool, with detailed options to
+specify behavior and 25 years of history dealing with weirdness.
+
+:code:`RSYNC` (also spelled :code:`NSYNC`) is a Charliecloud extension that
```
Comment (j-ogas) [inline doc/ch-image.rst:1034]: Can I really use NSYNC in my Dockerfile!? (love it).

**Review by j-ogas — CHANGES_REQUESTED** [review]
Few comments/questions in-line.

File: examples/Dockerfile.almalinux_8ch [hunk 2]
```diff
@@ -16,32 +16,15 @@ FROM almalinux:8
...
-                python3-sphinx \
-                python3-sphinx_rtd_theme \
-                rpm-build \
-                rpmlint \
-                rsync \
-                squashfs-tools \
-                squashfuse \
-                wget \
-                which \
+                autoconf automake gcc git libarchive libpng-devel make python3 python3-devel python3-lark-parser python3-requests python3-sphinx python3-sphinx_rtd_theme rpm-build rpmlint rsync squashfs-tools squashfuse wget which \
```
[the remark this answers is not in the document]
Comment (reidpr) [reply examples/Dockerfile.almalinux_8ch:27]: sadly yes 😭

File: lib/filesystem.py [hunk 3]
```diff
@@ -35,46 +36,64 @@
...
+
+   __slots__ = ("trailing_slash_p",)
 
    # Name of the gzip(1) to use; set on first call of file_gzip().
    gzip = None
 
+   def __init__(self, *args):
+      # WARNING: Something very weird is going on with the superclass [1] that
+      # means super().__init__ refers to object.__init__, not
+      # PosixPath.__init__. All seems well if we don’t call it but omg.
```
[the remark this answers is not in the document]
Comment (reidpr) [reply lib/filesystem.py:92]: Indeed. I'm thinking about changing `Path` so it no longer subclasses `pathlib.Path`.

File: doc/ch-image.rst (hunk 1 above)
Comment (reidpr) [reply doc/ch-image.rst:1034]: Yes you can!

File: lib/filesystem.py [hunk 4]
```diff
@@ -340,7 +365,7 @@ def is_relative_to(self, *other):
 
    def joinpath_posix(self, right):
       # This method is a hot spot, so the hairiness is due to optimizations.
-      # It runs about 30% faster than the naïve verson below.
+      # It runs about 30% faster than the naïve verson referenced below.
```
Comment (j-ogas) [inline lib/filesystem.py:368]:
```suggestion
      # It runs about 30% faster than the naïve version referenced below.
```

**Review by j-ogas — CHANGES_REQUESTED** [review]

**Review by j-ogas — APPROVED** [review]
LGTM

**Pull request merged**